include/pmacc/memory/boxes/PitchedBox.hpp

```diff
@@ -23,154 +23,195 @@
+        using RefValueType = ValueType&;
+
+        /** return value the origin pointer is pointing to
+         *
+         * @return value at the current location
+         */
+        HDINLINE RefValueType operator*()
+        {
+            return *this->m_ptr;
+        }
 
-    public:
-        using ReducedType = PitchedBox<TYPE, DIM2>;
+        /** get origin pointer
+         *
+         * @{
+         */
+        HDINLINE ValueType const* getPointer() const
+        {
+            return this->m_ptr;
+        }
+
+        HDINLINE ValueType* getPointer()
+        {
+            return this->m_ptr;
+        }
+        /** @} */
 
         /*Object must init by copy a valid instance*/
-        HDINLINE PitchedBox() = default;
+        PitchedBox() = default;
 
-        HDINLINE PitchedBox(TYPE* pointer, const size_t pitch, const size_t pitch2D)
-            : Base{pointer}
-            , pitch(pitch)
-            , pitch2D(pitch2D)
+        /** Constructor
+         *
+         * @tparam T_MemoryIdxType Index type
+         * @param pointer pointer to the memory
+         * @param extent extent of the memory in elements
+         * @param pitch line pitch in bytes
+         */
+        template<typename T_MemoryIdxType>
+        HDINLINE PitchedBox(
+            T_Type* pointer,
+            [[maybe_unused]] math::Vector<T_MemoryIdxType, DIM1> const& extent,
+            [[maybe_unused]] size_t const pitch)
+            : m_ptr(pointer)
         {
         }
 
-        /** constructor
+        HDINLINE PitchedBox(T_Type* pointer) : m_ptr(pointer)
+        {
+        }
+
+        PitchedBox(PitchedBox const&) = default;
+
+        /** get value at the given index
          *
-         * @param pointer pointer to the origin of the physical memory
-         * @param offset offset (in elements)
-         * @param memSize size of the physical memory (in elements)
-         * @param pitch number of bytes in one line (first dimension)
+         * @tparam T_MemoryIdxType Index type
+         * @param idx offset relative to the origin pointer
+         * @return reference to the value
+         * @{
          */
-        ///@todo(bgruber): is this functionality not provide by DataBox::shift?
-        HDINLINE PitchedBox(TYPE* pointer, const DataSpace<DIM3>& memSize, const size_t pitch)
-            : Base{pointer}
-            , pitch(pitch)
-            , pitch2D(memSize[1] * pitch)
+        HDINLINE T_Type const& operator[](int const idx) const
+        {
+            return *(m_ptr + idx);
+        }
+
+        HDINLINE T_Type& operator[](int const idx)
         {
+            return *(m_ptr + idx);
         }
 
-        HDINLINE PitchedBox(PitchedBox const&) = default;
+        HDINLINE T_Type const& operator[](DataSpace<DIM1> const& idx) const
+        {
+            return *(m_ptr + idx.x());
+        }
 
-        HDINLINE ReducedType operator[](const int idx) const
+        HDINLINE T_Type& operator[](DataSpace<DIM1> const& idx)
```

Comment:
Are ```const_cast```s required here and on line 197?

Reply:
Also a general styling comment. There is mixed use of T_Type and ValueType in the code, which are defined to be equal.

Reply:
`const_cast` should not be required.
I fixed the second point.